lib/src/source/hosted.dart

```diff
@@ -1100,3 +1109,91 @@ class _RefAndCache {
   @override
   bool operator ==(Object other) => other is _RefAndCache && other.ref == ref;
 }
+
+const checksumHeaderName = 'x-goog-hash';
+
+bool _hasChecksumHeader(Map<String, String> headers) {
+  return headers.containsKey(checksumHeaderName);
+}
+
+/// Adds a checksum validation "tap" to the response stream and returns a
+/// wrapped Stream object, which should be used to consume the incoming data.
+///
+/// As chunks are received, a CRC32C checksum is updated.
+/// Once the download is completed, the final checksum is compared with
+/// the one present in the checksum response header.
+///
+/// Throws [PackageIntegrityException] if anything is wrong with the checksum
+/// or if there is a checksum mismatch.
+Stream<List<int>> _responseStreamWithChecksumValidationTap(
```

Comment:
Probably easier to read this as an async* method. Something along the lines of:
```dart
Stream<List<int>> _validateChecksum(Stream<List<int>> stream, int? expectedChecksum) async* {
  if (checksum == null) yield* stream;
  final crc32 = Crc32();
  await for(final chunk in stream) {
    crc32.update(chunk);
  }
  final actualChecksum = crc32.finalize();
  if (actualChecksum != expectedChecksum) {
    // throw here.
  }
}
```

Reply:
Oh, yea that's a lot more compact and clear. I'll give it a shot.

Reply:
So, I tried this out and unfortunately, it ends up consuming the file stream entirely instead of tapping it.

Reply:
Actually, I forgot to add a `yield chunk;` link in the `await for` when I tried this yesterday. After adding that in, this approach works!